refactor(Build) Don't copy skeleton

michaelw85 opened the pull request:
Refactor & update based on PR comments by @3cp 
- Improved build time by keeping skeleton files as-is, they will not be copied to the dist folder.
- Added build step to pretest.
- Lib is now directly transpiled to dist instead of dist/lib
- Bin is no longer transpiled as it only serves as entrypoint

Related to #1067

michaelw85 commented:
@3cp I did a pre-release check as described in the release instructions. I solved any file related issue but ended up with `Error: Cannot find module 'gulp-sourcemaps'` not sure if this is due to my changes or my local setup. Since this is a package missing I would not expect this to be related to adding a transpiler.

3cp commented:
That was introduced by another small PR. Need a fix to add a condition in two files.

`cli-bundler/aurelia_project/tasks__if_typescript/transpile.ts`

```ts
// @if feat.plugin
import * as gulpSourcemaps from 'gulp-sourcemaps';
// @endif
```


`cli-bundler/aurelia_project/tasks__if_babel/transpile.js`

```ts
// @if feat.plugin
import gulpSourcemaps from 'gulp-sourcemaps';
// @endif
```

3cp commented:
Your missing detection of local aurelia-cli installation is due to release check install aurelia-cli using github repo `michaelw85/cli#refactor_build` which didn't trigger a build.

It cannot find `dist/index.js` file.

You need two fixes:
1. turn on `"prepare": "npm run build"` in package.json, that will ensure build when using github repo, and ensure build before `npm publish`.
2. Change Yarn to NPM in two files, because yarn has a bug that bypasses `"prepare"` script.
`build/tasks/release-checks/tasks/install-latest-aurelia-cli.js`
`build/tasks/release-checks/tasks/install-node-modules.js`

michaelw85 commented:
Thx for the info. Should these changes be part of this PR? Or just local to be able to do the release check?

3cp commented:
release-check is part of the git repo. Yes, please commit all needed.

michaelw85 commented:
Ok will update the pr asap, I might be able to spend time tonight again.

michaelw85 commented:
PR is updated and did a release-check after. It passed for the first 2 skeletons, looked promising and then:
```cmd
DEBUG [20:04:23]  Verifying Cypress can run C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress [started]
DEBUG [20:04:24]  Verifying Cypress can run C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress [failed]
[20:04:24] '<anonymous>' errored after 4.78 min
[20:04:24] Error: Test failed
    at AuCypressRunsTests.fail (C:\repos\aurelia\cli\build\tasks\release-checks\tests\test.js:24:11)
    at AuCypressRunsTests.onCypressOutput (C:\repos\aurelia\cli\build\tasks\release-checks\tests\generic\au-cypress.js:17:12)
    at ExecuteCommand.cypressCommand.ExecuteCommand [as outputCallback] (C:\repos\aurelia\cli\build\tasks\release-checks\tests\generic\au-cypress.js:27:90)
    at Socket.proc.stdout.on (C:\repos\aurelia\cli\build\tasks\release-checks\tasks\execute-command.js:32:12)
    at Socket.emit (events.js:182:13)
    at Socket.EventEmitter.emit (domain.js:459:23)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at Pipe.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
[20:04:24] 'release-check' errored after
```
I don't have time to investigate this issue right now, but it seems to originate from the following skeleton:
`cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum`

3cp commented:
That's the first cypress skeleton you encountered. It might be some cypress config need to be updated.

BTW, I tried to test your branch, I still got the missing json in dist/ folder.
Because the missing quotes in your npm scripts, you need that, otherwise, some unix shell will replace those wildcard before openfiles sees the args.

```
"build:copy-files": "copyfiles -u 1 -e \"./lib/**/*.{js,ts}\" \"./lib/**/*\" ./dist",
```

These added quotes (pls use double-quote, as they are the only one works on win32) should work for all OS.

I am having a look on the cypress.

3cp commented:
On the exact skeleton, the cypress and all other tests passed on my mac.
I am not sure what went wrong on your PC.

3cp commented:
FYI, all release-check passed on my mac. BTW, it's very slow now because of NPM...

3cp commented on package.json:
```diff
@@ -24,17 +24,18 @@
     "build:ts:watch": "npm run build:ts -- --watch",
     "build:dts": "tsc -d --emitDeclarationOnly --allowJs false",
-    "build:copy-files": "copyfiles -e ./skeleton/**/*.{js,ts} -e ./lib/**/*.{js,ts} ./lib/**/* ./skeleton/**/* ./dist",
+    "build:copy-files": "copyfiles -u 1 -e ./lib/**/*.{js,ts} ./lib/**/* ./dist",
```
Please add necessary quotes
```
"build:copy-files": "copyfiles -u 1 -e \"./lib/**/*.{js,ts}\" \"./lib/**/*\" ./dist",
```

michaelw85 replied:
Done

michaelw85 commented:
FYI here's the debug log on the failing cypress test:

```
Platform: win32 (10.0.17763)
Cypress Version: 3.4.0
    at C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\cypress\lib\errors.js:202:15
    at bound (domain.js:395:14)
    at runBound (domain.js:408:12)
    at tryCatcher (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:512:31)
    at Promise._settlePromise (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:569:18)
    at Promise._settlePromise0 (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:614:10)
    at Promise._settlePromises (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:693:18)
    at Async._drainQueue (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:133:16)
    at Async._drainQueues (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:143:10)
    at Immediate.Async.drainQueues [as _onImmediate] (C:\repos\aurelia\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:17:14)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23) known: true, context: [Object: null prototype] {} }
(node:7728) [DEP0097] DeprecationWarning: Using a domain property in MakeCallback is deprecated. Use the async_context variant of MakeCallback or the AsyncResource class instead.
```
Seems related to my bluebird version for some reason, not sure if i'm behind or ahead.

3cp commented:
Our Bluebird was replaced by promise-polyfill few months ago. Your bluebird is a dep of cypress. Something messed up in your local cypress installation. Maybe try a flesh nodejs version to get clean env.

```
⋊> ~/p/s/cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum 
npm list bluebird
cli-bundler-alameda-dotnet-core-babel-jest-cypress-scaffold-minimum@0.1.0 /Users/huocp/playground/ske/cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum
└─┬ cypress@3.4.0
  └── bluebird@3.5.0 
```

3cp commented:
I am testing your branch on a win10 box, I am getting very weird result.

Win10, node v10.15.2, npm v6.10.2.

Fresh clone, fresh `npm i`, then `npm run build:ts`.

It didn't generate `dist/index.js` as expected, it generated `dist/cli/lib/index.js`. There is another unexpected file `dist/enquirer/lib/utils.js` generated. I have no idea what's going on with tsc.

michaelw85 commented:
hmm that is unexpected, maybe I should scope the include to the lib folder for tsc

michaelw85 commented:
I checked the tsconfig and I already scope it specific to lib, I have no clue how it would end up with anything else than lib content on your machine. Stupid question but did you run it from the correct dir?

```yml
  "include": [
    "./lib/**/*"
  ]
```

This is the result of a clean build on my machine:
![image](https://user-images.githubusercontent.com/3629094/61990787-4814cc80-b047-11e9-9e03-f3b369b4a546.png)

3cp commented:
Has to be in right folder, otherwise npm script would not run. I will have an another look on Monday.

Does tsc support some shared/global config that might mess up?

michaelw85 commented:
O right, as I said stupid question...
I'm not aware of an option to automatically pick up any global config, you can extend other configs but this is an explicit action from the config

michaelw85 commented:
Ok tried again, removed and reinstalled node LTS. Removed node_modules, npm i + run build.
Seemed to start pretty well, it passed the earlier failing cypress test but after a while unfortunately crashed again.

`cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum`
```
DEBUG [12:08:57]  Verifying Cypress can run C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress [started]
DEBUG [12:08:58]  Verifying Cypress can run C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress [failed]
[12:08:58] '<anonymous>' errored after 3.65 min
[12:08:58] Error: Test failed
    at AuCypressRunsTests.fail (C:\repos\aurelia\cli\build\tasks\release-checks\tests\test.js:24:11)
    at AuCypressRunsTests.onCypressOutput (C:\repos\aurelia\cli\build\tasks\release-checks\tests\generic\au-cypress.js:17:12)
    at ExecuteCommand.cypressCommand.ExecuteCommand [as outputCallback] (C:\repos\aurelia\cli\build\tasks\release-checks\tests\generic\au-cypress.js:27:90)
    at Socket.proc.stdout.on (C:\repos\aurelia\cli\build\tasks\release-checks\tasks\execute-command.js:32:12)
    at Socket.emit (events.js:198:13)
    at Socket.EventEmitter.emit (domain.js:466:23)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at Pipe.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
[12:08:58] 'release-check' errored after
```
Manually attempted to run:

```
C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum> au cypress
Local aurelia-cli v1.0.2
Starting '_default'...
It looks like this is your first time using Cypress: 3.4.0

 ×  Verifying Cypress can run C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress
   → Cypress Version: 3.4.0
Uncaught promise rejection:
{ Error: Cypress failed to start.

This is usually caused by a missing library or dependency.

The error below should indicate which dependency is missing.

https://on.cypress.io/required-dependencies

If you are using Docker, we provide containers with all required dependencies installed.

----------

Command failed: C:\Users\micha\AppData\Local\Cypress\Cache\3.4.0\Cypress\Cypress.exe --smoke-test --ping=667

----------

Platform: win32 (10.0.18362)
Cypress Version: 3.4.0
    at C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\cypress\lib\errors.js:202:15
    at bound (domain.js:402:14)
    at runBound (domain.js:415:12)
    at tryCatcher (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:512:31)
    at Promise._settlePromise (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:569:18)
    at Promise._settlePromise0 (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:614:10)
    at Promise._settlePromises (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\promise.js:693:18)
    at Async._drainQueue (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:133:16)
    at Async._drainQueues (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:143:10)
    at Immediate.Async.drainQueues [as _onImmediate] (C:\repos\sk\cli-bundler_alameda_dotnet-core_babel_jest_cypress_scaffold-minimum\node_modules\bluebird\js\release\async.js:17:14)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:126:23) known: true, context: [Object: null prototype] {} }
(node:8684) [DEP0097] DeprecationWarning: Using a domain property in MakeCallback is deprecated. Use the async_context variant of MakeCallback or the AsyncResource class instead.
```

I'm getting a bit annoyed with these failures on the pre-release check. Did these ever successfully run on a windows machine? How likely is it this is related to the build changes?

3cp commented:
I ran the release-check many times before on win10 (I mean before TS PR), never saw cypress fail like that. This is very strange.

michaelw85 commented:
I'm going to remove some cache and try again, just noticed the cache dir in de logs. Not sure what to attempt after that, I could try my desktop to see what happens.

michaelw85 commented:
Currently looking promising so far: `Executing 16/72` it's taking a long time as you said but it seems to be running after removing my cache. I guess my cypress cache has some deprecated version/files

michaelw85 commented:
It finally finished running after move than 3 hours but it succeeded!
[summary.txt](https://github.com/aurelia/cli/files/3439230/summary.txt)

3cp commented:
Hey, I found out how to reproduce my weird tsc issue. It looks like tsc has some stupid bug if this is not unawarely intended.

If your cli repo is in location `C:\some\path\cli`, do `git clone git@github.com:enquirer/enquirer.git` in `C:\some\path\` so that enquirer repo is a sibling folder `C:\some\path\enquirer`.

Now goto folder `C:\some\path\cli` and cleanup `dist/` folder, then run `npm run build:ts`, see the strange `dist/` output in cli repo.

```
dist/cli/lib/... (instead of just dist/...)
dist/enquirer/lib/utils
```

Some TS user please tell me what's going on. This is so weird. cc @bigopon

3cp commented:
@michaelw85 good news. I now know how to bypass this issue.

Add `"module": "commonjs",` to `tsconfig.json` compilerOptions. It's now compiled as expected!

The default `module` is `es6` when `"target":"es6"` is used. You can reproduce the issue by explicitly set `"module": "es6",` in `tsconfig.json`.

I don't understand why the module format affected the output.

Anyway, let's fix it and merge!

3cp commented:
I got bit more understanding on the issue.

This is caused by `lib/string.js` `const {wordWrap} = require('enquirer/lib/utils');`.
This is a nodejs require which TypeScript default module resolution doesn't understand. So the issue can be fixed by adding `"moduleResolution": "node"` to tsconfig even without `"module": "commonjs"`.

I have added both to our tsconfig, because this is a Nodejs module, both options are required logically.

```
"module": "commonjs",
"moduleResolution": "node"
```

FYI, the strange behaviour is defined in TypeScript default module resolution "Classic". It keeps looking into parent folders, it just happened there is an `enquirer` folder in its search path.
https://www.typescriptlang.org/docs/handbook/module-resolution.html#classic
